TOOLS-1648 Add function for only counting diffs

razvan-pricop-ptt opened the pull request:
## Overview

Three functions were added, only one of the being important. The other two were added to keep the source code consinstent.
The new functions:
- `diff_main_only_count` - analogous to `diff_main`(line 271) and `diff_main`(line 326) combined
- `diff_compute_only_count` - analogous to `diff_compute`
- `diff_bisect_only_count` - the function that counts the diffs. Its name was kept for historic reason, but it doesn't perform any bisection.

The counting algorithm is simpler than the original `diff_bisect`, which was recursive and performed path finding starting from two sources in parallel.

## The implementation of `diff_bisect_only_count`

The algorithm is taken from [Myers' paper](https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.4.6927), as all the sources suggest. For a visual and interactive reference [check the article from RobertElder](https://blog.robertelder.org/diff-algorithm/). In essence, it is a shortest path search in a graph defined by them, in which they perform an algorithm similar to BFS in principle, but different in construction. For storing the current state it uses just an array.

We started from the classic Myers algorithm, and added a few modifications. We borrowed one optimization from `diff_bisect` which defines `kstart` and `kend`. These are used to prune the iteration of the diagonals. From personal experimentation, additional optimizations are relevant only if we care about where are the differences. But in our case, we only care about how many of them there are.

In addition to the algorithm, we added a max diff limit argument. The function starts looking for differences starting from 0, then 1, 2, 3, 4, ... We chose to give the option to stop at max diff. That's because. in practice, if our HTML are more than 4% different, we consider them distinct.

## Testing

For automatic testing, run the following 4 commands:
```
g++ diff_match_patch_test_string.cpp -o diff_match_patch_test && ./diff_match_patch_test
g++ -O3 diff_match_patch_test_string.cpp -o diff_match_patch_test && ./diff_match_patch_test
g++ diff_match_patch_test_wstring.cpp -o diff_match_patch_test && ./diff_match_patch_test
g++ -O3 diff_match_patch_test_wstring.cpp -o diff_match_patch_test && ./diff_match_patch_test
```

iulian-birlica commented on diff_match_patch.h:
```diff
@@ -281,6 +281,35 @@ class diff_match_patch {
+      deadline = clock() + (clock_t)(Diff_Timeout * CLOCKS_PER_SEC);
+    }
+
+    long long diff_count = 0;
```
nit: This seems like it can also be `unsigned long long`. That would, of course, necessitate changes in the return types.

iulian-birlica commented on diff_match_patch.h:
```diff
@@ -404,6 +433,37 @@ class diff_match_patch {
+    }
+
+    {
+      const string_t& longtext = text1.length() > text2.length() ? text1 : text2;
```
nit: The comparison could be cached, but I assume the compiler will do that at higher optimization flags.

razvan-pricop-ptt replied:
This was copied from the previous source.

iulian-birlica commented on diff_match_patch.h:
```diff
@@ -404,6 +433,37 @@ class diff_match_patch {
+    {
+      const string_t& longtext = text1.length() > text2.length() ? text1 : text2;
+      const string_t& shorttext = text1.length() > text2.length() ? text2 : text1;
+      const size_t i = longtext.find(shorttext);
```
note: Not sure this is a speedup, if most of the strings are not contained in each other, but it seems benign enough

razvan-pricop-ptt replied:
Again, taken from the previous source. I also think this is not a significant improvement.

iulian-birlica commented on diff_match_patch.h:
```diff
@@ -590,6 +650,99 @@ class diff_match_patch {
+      v1[i] = 0;
+    }
+
+    // Solve for d == 0
```
note: nice dynamic programming problem.

iulian-birlica reviewed (approved):
Seems really nice, maybe overflow prone, but that would mean having terabytes of strings, so not really a problem.

iulian-birlica commented:
Do you happen to know what the speedup was, on the original test that triggered the entire discussion about making this faster? Or any other benchmark?

stefan-mihalache commented on diff_match_patch_test_string.cpp:
```diff
@@ -638,6 +639,70 @@ class diff_match_patch_test : diff_match_patch<string> {
+    assertEquals("diff_main_only_count: Different.", 6, dmp.diff_main_only_count("abc", "def", 0));
+    assertEquals("diff_main_only_count: Overlap #2.", 6, dmp.diff_main_only_count("abcy", "xaxcxabc", 6));
+    assertEquals("diff_main_only_count: Overlap #2.", 6, dmp.diff_main_only_count("abcy", "xaxcxabc", 7));
+    assertEquals("diff_main_only_count: Overlap #2.", 12, dmp.diff_main_only_count("abcy", "xaxcxabc", 5));
```
How can this be 12 when max_diff is given as 5?

razvan-pricop-ptt replied:
This translates to: "Check if diff doesn't exceed a limit, otherwise consider the texts completely different". In our case, 12 is the maximum possible difference. The function couldn't return 5, because that would mean the diff length is 5, which is false. So it returns the worst case.

The same thing happens for the diff in website_scanner, when the timeout is exceeded.

stefan-mihalache commented on diff_match_patch.h:
```diff
@@ -404,6 +433,37 @@ class diff_match_patch {
+      if (shorttext.length() == 1) {
+        // Single character string.
+        // After the previous speedup, the character can't be an equality.
+        return 1 + longtext.size();
```
Nit: I see that you use both `.length()` and `.size()`. Why not just pick one of them to use instead? (size is more general in C++)

razvan-pricop-ptt replied:
Chose length because it's more frequent in the file.

stefan-mihalache commented on diff_match_patch.h:
```diff
@@ -590,6 +650,99 @@ class diff_match_patch {
+      while (text1[x] == text2[y]) {
+        x++;
+        y++;
+      }
```
isn't this the common prefix part that has already been checked in one of the previous functions?

razvan-pricop-ptt replied:
Yes, it is. Nice spotted!

stefan-mihalache commented on diff_match_patch.h:
```diff
@@ -590,6 +650,99 @@ class diff_match_patch {
+    }
+
+    int d_max = text1_len + text2_len;
+    int len_min = text1_len;
```
how can you be sure that text1 has the shortest length? Also,`len_min` and `len_max` don't seem to be used anywhere else in this function.

razvan-pricop-ptt replied:
Yes, that should be deleted.

stefan-mihalache reviewed (approved):
The code seems ok overall. 

Still curios about some benchmark results/numbers once they are done.

razvan-pricop-ptt committed:
feat: add function for only counting diffs

razvan-pricop-ptt committed:
refactor: undo changes applied to initial diff function

razvan-pricop-ptt committed:
doc: add comments explaining diff algorithm

razvan-pricop-ptt committed:
refactor: remove typedef for ints

razvan-pricop-ptt committed:
test: count function

razvan-pricop-ptt committed:
feat: remove checklines argument from diff function

razvan-pricop-ptt committed:
feat: increase buffer size for possible overflow

razvan-pricop-ptt committed:
refactor: replace size() with length()

razvan-pricop-ptt committed:
refactor: remove unnecesary code

Remove the code for the special case when d = 0. It's already treated when we remove the prefix.
Remove len_min and len_max unused variables.

razvan-pricop-ptt commented:
> Do you happen to know what the speedup was, on the original test that triggered the entire discussion about making this faster? Or any other benchmark?

Check other PRs.

stefan-mihalache reviewed (approved):
Looks good to go now. Nice speedup.
